File: backend/src/ytforge/interfaces/api/schemas/videos.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from ytforge.domain.enums import VideoStatus


class VideoCreateRequest(BaseModel):
    render_asset_id: uuid.UUID
    title: str = Field(min_length=1, max_length=100)
    description: str
    synthetic_content_disclosure: bool = True


class VideoUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = None


class VideoRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    render_asset_id: uuid.UUID
    title: str
    description: str
    status: VideoStatus
    synthetic_content_disclosure: bool
    youtube_video_id: str | None
    scheduled_publish_at: datetime | None
    published_at: datetime | None


class SeoMetadataSetRequest(BaseModel):
    title: str = Field(min_length=1, max_length=100)
    description: str = Field(min_length=1, max_length=5000)
    thumbnail_asset_id: uuid.UUID | None = None
    tags: list[Any] = Field(default_factory=list)
    chapters: list[Any] = Field(default_factory=list)
    keywords: list[Any] = Field(default_factory=list)


class SeoMetadataRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    video_id: uuid.UUID
    title: str
    description: str
    thumbnail_asset_id: uuid.UUID | None
    tags: list[Any]
    chapters: list[Any]
    keywords: list[Any]

File: frontend/src/lib/api/schemas/videos.ts
import { z } from "zod";
import { VideoStatus } from "@/types/enums";

const videoStatusSchema = z.enum([
  VideoStatus.DRAFT,
  VideoStatus.UPLOADED,
  VideoStatus.SCHEDULED,
  VideoStatus.PUBLISHED,
  VideoStatus.FAILED,
]);

export const videoReadSchema = z.object({
  id: z.string().uuid(),
  project_id: z.string().uuid(),
  render_asset_id: z.string().uuid(),
  title: z.string(),
  description: z.string(),
  status: videoStatusSchema,
  synthetic_content_disclosure: z.boolean(),
  youtube_video_id: z.string().nullable(),
  scheduled_publish_at: z.string().nullable(),
  published_at: z.string().nullable(),
});
export type VideoRead = z.infer<typeof videoReadSchema>;

export const seoMetadataReadSchema = z.object({
  id: z.string().uuid(),
  video_id: z.string().uuid(),
  title: z.string(),
  description: z.string(),
  thumbnail_asset_id: z.string().uuid().nullable(),
  tags: z.array(z.unknown()),
  chapters: z.array(z.unknown()),
  keywords: z.array(z.unknown()),
});
export type SeoMetadataRead = z.infer<typeof seoMetadataReadSchema>;

export type VideoCreateRequest = {
  render_asset_id: string;
  title: string;
  description: string;
  synthetic_content_disclosure?: boolean;
};

export type VideoUpdateRequest = {
  title?: string | null;
  description?: string | null;
};

export type SeoMetadataSetRequest = {
  title: string;
  description: string;
  thumbnail_asset_id?: string | null;
  tags?: unknown[];
  chapters?: unknown[];
  keywords?: unknown[];
};
